Add System.Runtime.CompilerServices.Unsafe library, tests, and packaging

mellinoe opened the pull request:
This adds a preliminary version of `System.Runtime.CompilerServices.Unsafe`, with tests and packaging. Included are four basic methods for manipulating pointers generically:

````CSharp
public static class Unsafe
{
    public unsafe static T Read<T>(void* p);
    public unsafe static void Write<T>(void* p, T value);
    public unsafe static void* AddressOf<T>(ref T value);
    public static int SizeOf<T>();
}
````

This library addresses special situations where it is desirable to manipulate types generically using pointers. In C#, these scenarios are very poorly supported (or rather, not at all), but IL has richer support for operations like this. The library is implemented in IL, and utilizes a "ilproj" file to participate in the build, which is supported by the newest version of buildtools. The test project is a regular C# test library consuming the built assembly from the IL project.

The package has minimal dependencies (System.Runtime v4.0.0.0) and therefore can be installed all the way down to netstandard1.0.

NOTE: This can't be built outside of Windows, yet. It seems the nuget packages for ILAsm on other platforms have some issues at this time. I have manually verified that the build will work when we are able to install ILAsm from a NuGet package on those platforms, though.

@jkotas, @terrajobst, @stephentoub, @nietras

Addresses https://github.com/dotnet/corefx/issues/5474

stephentoub commented on src/System.Unsafe/src/System.Unsafe.il:
```diff
@@ -0,0 +1,91 @@
+
+// =============== CLASS MEMBERS DECLARATION ===================
+
+.class public abstract auto ansi sealed beforefieldinit System.Unsafe
```
Could you include a manually generated .XML doc file that could be included in the NuGet package to provide descriptive IntelliSense?  You could have the compiler create one for you if you created a dummy C# project with the same APIs, wrote out the XML comments you wanted, had the compiler generate the doc file for you, etc.

jkotas replied:
I think it should be in some more specific namespace. (`System.Runtime.CompilerServices` would be my candidate.)

justinvp replied:
Also, the assembly name would ideally align with the namespace, e.g. `System.Runtime.CompilerServices.Unsafe`

mellinoe replied:
That seems like a good namespace for it, I'll switch everything to use that name (library, package, class)

mellinoe replied:
I've changed the name of everything in the second commit.

NOTE: Still need to add XML documentation.

jkotas commented on src/System.Unsafe/src/System.Unsafe.il:
```diff
@@ -0,0 +1,91 @@
+.class public abstract auto ansi sealed beforefieldinit System.Unsafe
+       extends [System.Runtime]System.Object
+{
+  .method public hidebysig static !!T  Read<T>(void* p) cil managed aggressiveinlining
```
Could you please augment these with  NonVersionable attribute to make these inlined even for R2R?

(Here is an example on how to do it: https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601.cs#L120)

mellinoe replied:
I added this. I'm assuming that the type is matched by name, since Immutable Collections are defining their own internal type?

jkotas replied:
Right.

stephentoub commented on src/System.Unsafe/pkg/System.Unsafe.builds:
```diff
@@ -0,0 +1,9 @@
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <ItemGroup>
+    <Project Include="System.Unsafe.pkgproj" Condition="'$(OS)'=='Windows_NT'" />
```
Should we have an issue for the Windows_NT only issue and include the issue # here in a comment?

jkotas commented on src/System.Unsafe/src/System.Unsafe.il:
```diff
@@ -0,0 +1,91 @@
+  .method public hidebysig static void*  AddressOf<T>(!!T& 'value') cil managed aggressiveinlining
+  {
+    .maxstack  8
+    IL_0000:  ldarg.0
```
Delete the `IL_0000` labels for consistency

jkotas commented on src/System.Unsafe/src/System.Unsafe.il:
```diff
@@ -0,0 +1,91 @@
+
+  .method public hidebysig static void*  AddressOf<T>(!!T& 'value') cil managed aggressiveinlining
+  {
+    .maxstack  8
```
Make the indentation consistent with above.

jkotas commented on src/System.Unsafe/src/System.Unsafe.il:
```diff
@@ -0,0 +1,91 @@
+  {
+    .maxstack  8
+    IL_0000:  ldarg.0
+    IL_0002:  ret
```
The result needs to be explicit converted using `conv.u` to avoid confusing the JIT - as discussed in the issue.

jkotas commented on src/System.Unsafe/src/System.Unsafe.il:
```diff
@@ -0,0 +1,91 @@
+          ret
+  } // end of method Unsafe::Write
+
+  .method public hidebysig static void*  AddressOf<T>(!!T& 'value') cil managed aggressiveinlining
```
`AsPointer` may be a better name - as discussed in the issue.

jkotas replied:
Also, it would be really nice to have As<class T>, InitBlock, CopyBlock and Pinnable - as discussed in the issue.

mellinoe replied:
I will follow up with those four extra methods.

mellinoe commented:
More changes based on feedback:

* Added `Unsafe.CopyBlock`, `Unsafe.InitBlock`, `Unsafe.As<T>()`
* Added XML documentation

justinvp commented on pkg/descriptions.json:
```diff
@@ -1654,6 +1654,13 @@
         ]
     },
     {
+        "Name": "System.Runtime.CompilerServices.Unsafe",
```
Nit: Are the entries in this file sorted?

mellinoe replied:
Ahh good point. They are (mostly) sorted, and this is no longer in the right place.

justinvp commented on src/System.Runtime.CompilerServices.Unsafe/tests/UnsafeTests.cs:
```diff
@@ -0,0 +1,162 @@
+using System.Runtime.InteropServices;
+using Xunit;
+
+namespace System.Runtime.CompilerServices
```
Nit: Should the namespace be `System.Runtime.CompilerServices.Tests`?

mellinoe replied:
I'm not sure we have a guideline for this, I know it's a bit inconsistent elsewhere in the repo. I sort of prefer leaving off "Tests" since the class names almost always have "Tests" in them, but that might just be me.

jkotas commented:
Some tests for the new methods (As, InitBlock, CopyBlock) would be nice. LGTM otherwise.

mellinoe commented:
I've added tests for `As`, `InitBlock`, and `CopyBlock`.

@jkotas Are there additional interesting test cases you think I should add for `As`, `Read`, and/or `Write`? I can think of some additional cases, but I am not sure what is safe or valid to be testing.

jkotas commented:
I do not think it needs extensive coverage for various cases - it is just wrapping a single IL instruction that we should have coverage via codegen tests.

mellinoe commented:
Sounds reasonable to me. The coverage here will let us know that the library has build successfully and is usable in applicable scenarios, but lower-level CoreCLR tests will cover the behavior more thoroughly.

jkotas commented on src/System.Runtime.CompilerServices.Unsafe/tests/UnsafeTests.cs:
```diff
@@ -0,0 +1,236 @@
+        public static void As()
+        {
+            object o = "Hello";
+            Assert.Equal("Hello", Unsafe.As<string>(o));
```
```
[Fact]
public static void DangerousAs()
{
    // Verify that As does not perform type checks
    object o = new Object();
    Assert.IsType(typeof(Object), Unsafe.As<string>(o));
}
```

jkotas replied:
This is interesting case that I would add.

mellinoe replied:
Added.

jamesqo commented:
@mellinoe Not from Microsoft but, maybe we could add IntPtr-based overloads that would allow the methods to be called from safe code (including languages that don't support unsafe, like VB)?

jkotas commented:
@jamesqo It is by design that these unsafe APIs are designed to be only usable from languages that allow writing unsafe code already. It is the general principle that we are following for any new unsafe APIs - e.g. `Buffer.MemoryCopy`.

mellinoe commented:
I realized that the runtimes section was missing from my test's project.json. I've added it, and hopefully this fixes the CentOS test failures (it wasn't able to find some DLL for some reason, not sure which or why).

mellinoe commented:
@dotnet-bot Test Innerloop CentOS7.1 Debug Build and Test

https://github.com/dotnet/corefx/issues/7805

mellinoe commented:
@dotnet-bot Test Innerloop CentOS7.1 Debug Build and Test

Hit the same error again...

davidfowl commented:
> The package has minimal dependencies (System.Runtime v4.0.0.0) and therefore can be installed all the way down to netstandard1.0.

You don't need to target the lowest version of he packages to support the lowest netstandard. In fact, the versions of those packages don't even have the correct target frameworks. You should be targeting the latest versions of the packages

mellinoe commented:
> You don't need to target the lowest version of he packages to support the lowest netstandard

I didn't mean to imply I'm using the lowest version of the packages to build this. Actually, I'm not using any packages to build this, it's an IL project.

nietras commented:
@mellinoe you could just have started the implementation of this by copying https://github.com/DotNetCross/Memory.Unsafe That is kind of why I open sourced it in the first place. Anyhow, it is great this is finally being added. I have a few questions though, mostly for clarification for me, so I hope you all will indulge me.

 * What about the `Pinnable` class? Can't see it in the commits. See https://github.com/DotNetCross/Memory.Unsafe/blob/master/il/DotNetCross.Memory.Unsafe.il#L130
 * Why `corefx`? And not `coreclr`? As I understood it there where several use cases for this in `coreclr`. Although, I think it is great if this will be published as a separate package, so I guess it may still be used in `coreclr`. Or perhaps I misunderstood @jkotas when he mentioned internal use cases.
 * Any reason **not** to be specific about `.maxstack`? That is, why default to `8` and not say `1` when applicable?
 * There seems to be a lack of consistency in naming e.g. sometimes pointer is referred to as `p` other times it is specifically named `destination`, wouldn't it be preferable to name these `destination`, `source` where applicable? E.g. `Write<T>(void* destination, T!! 'value')`
 * Since C# 7 is out as a preview in VS "15" and with this local refs and ref returns (https://github.com/dotnet/roslyn/issues/118) would it not be useful to add the `AsRef<T>` method at the same time? This would mean this would be available from the get go and this is something we could really use a lot in the numerical processing we do, mainly due to issues with how the JIT handles compounded value types as detailed in https://github.com/dotnet/coreclr/issues/3539 (since we target .NET 4.0 in some cases we can't rely on future JIT improvements to help us). And as I understand it this can be defined in IL so we don't need C# 7 to make the feature, although testing it from C# won't be possible before I guess. If this was available now we could experiment with this feature.

nietras commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+  .method public hidebysig specialname rtspecialname 
+          instance void  .ctor() cil managed
+  {
+        // Code size       7 (0x7)
```
Remove code size comment.

jkotas commented:
> Why  corefx ? And not  coreclr ?

corefx repo is where we build all framework nuget packages, that may possibly go earlier runtimes. coreclr repo does not have the corefx infrastructure for building, testing, packaging and publishing these.

Moving it to coreclr would not actually solve the problem with internal mscorlib use. mscorlib cannot depend on anything else, and so the version for internal mscorlib use has to be done in special way anyway - it cannot just use the standalone version as everybody else.

> What about the  Pinnable  class?
> useful to add the  AsRef<T>

I agree that these would be pretty useful.

tmds commented:
Marshal.SizeOf has a few overloads. I find the one that accepts a Type argument handy, it does give me a warning though: 'Marshal.SizeOf(Type)' is obsolete.
Perhaps the Unsafe class can include the overloads that make sense?

mellinoe commented:
> @mellinoe you could just have started the implementation and basic tests of this by copying https://github.com/DotNetCross/Memory.Unsafe That is kind of why I open sourced it in the first place. 

Thanks, I will take a look at your tests and see if I can augment the ones here with some of yours. Which do you think would be the most valuable to add? As for the "library code", most of the work here was actually just getting things integrated into the build system, which wasn't difficult but needed some changes in our buildtools repo to work. It looks like our implementations are same at this point, modulo things like namespace, parameter names, etc.

> What about the Pinnable class? Can't see it in the commits. See https://github.com/DotNetCross/Memory.Unsafe/blob/master/il/DotNetCross.Memory.Unsafe.il#L130

This does seem to be useful in some cases, but I wonder if we actually need to implement it in the library. Obviously it would be useful to have it at your disposal with the rest of the library, but  you could also easily write it yourself, since it doesn't require any IL re-writing... I'm a bit conflicted but we can probably go ahead and add it.

> Any reason not to be specific about .maxstack? That is, why default to 8 and not say 1 when applicable?

Yes, I should change these. I actually had to remind myself what these were for (haven't done a lot of raw IL writing...). It seems like leaving the default of 8 isn't correct.

> There seems to be a lack of consistency in naming e.g. sometimes pointer is referred to as p other times it is specifically named destination, wouldn't it be preferable to name these destination, source where applicable? E.g. Write<T>(void* destination, T!! 'value')

I tried to use `p` when the pointer wasn't really a "destination" or "source". For `Cast()` or `InitBlock()`, you don't really have a source or destination, you just have a pointer. We don't have good guidelines for pointers in the BCL because we don't use a lot of pointers.

> There are no tests that actually show calls are inlined, since we had the conv.u issue would it not be beneficial to ensure these calls can in fact be inlined?

I'm happy to add extra tests to cover things that are missing. I'm not sure how we'd test this, though, suggestions?

As for adding an overload for the C#7 feature, I'm leaning towards waiting on that for now. We don't generally rely on any language features before the compiler is actually stable, and it may be a little while after that before corefx itself can actually rely on that compiler anyways. In this case, the overload wouldn't actually be usable from C# until that point.

> Marshal.SizeOf has a few overloads. I find the one that accepts a Type argument handy, it does give me a warning though: 'Marshal.SizeOf(Type)' is obsolete.
Perhaps the Unsafe class can include the overloads that make sense?

`Unsafe` is for exposing operations available in IL, and I don't think this falls into that bucket. I have felt this annoyance before, though, and I've thought it a bit odd that the method is obsolete. Could you file an issue regarding that?

jkotas commented:
> It seems like leaving the default of 8 isn't correct.

The default 8 is correct, it is just not precise. There is a special case for maxstack 8 in IL binary format that results in smaller encoding. It is why you often see in IL generated by high-level compilers - to make fine binary smaller. This size difference does matter in this case because of it is just a few method - using the precise maxstack is fine.

> `Unsafe` is for exposing operations available in IL, and I don't think this falls into that bucket.

Right, `Marshal.SizeOf`, and `Unsafe.SizeOf' are different:
- `Unsafe.SizeOf` returns the size of the managed representation of the type. It returns the same value as `sizeof` in C#, and it is about as fast.
- `Marshal.SizeOf` returns the size of the structure that the managed/unmanaged interop marshalls the type to. It may or may not be the same as `sizeof` in C#. Also, it is orders of magnitude slower than `sizeof` in C#.

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+  } // end of method Unsafe::Write
+
+  .method public hidebysig static void  Write<T>(void* p,
+                                                 !!T& 'value') cil managed aggressiveinlining
```
Just a question: why aren't there any tests for this overload of `Write`? I believe this is the equivalent of `Write<T>(void* p, ref T value)` in C#, right?

jkotas replied:
Is this overload really needed? It functionally equivalent to the other Write overload. I do not think it should be here.

I understand that one or the other may give better codegen in some cases depending on how the JIT is implemented, but having both makes it hard for folks to decide which one to use.

nietras replied:
Yes, the overload is very much needed. E. g. for compounded value types.
codegen is extremely important to us, so we need both versions. `Unsafe`
use is for advanced users anyway so I don't see the problem.
On Apr 25, 2016 03:57, "Jan Kotas" <notifications@github.com> wrote:

> In
> src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il
> <https://github.com/dotnet/corefx/pull/7966#discussion_r60857131>:
>
> > +        ret
> > +  } // end of method Unsafe::Read
> > +
> > +  .method public hidebysig static void  Write<T>(void* p,
> > +                                                 !!T 'value') cil managed aggressiveinlining
> > +  {
> > +        .custom instance void System.Runtime.Versioning.NonVersionableAttribute::.ctor() = ( 01 00 00 00 )
> > +        .maxstack  8
> > +        ldarg.0
> > +        ldarg.1
> > +        stobj !!T
> > +        ret
> > +  } // end of method Unsafe::Write
> > +
> > +  .method public hidebysig static void  Write<T>(void* p,
> > +                                                 !!T& 'value') cil managed aggressiveinlining
>
> Is this overload really needed? It functionally equivalent to the other
> Write overload. I do not think it should be here.
>
> I understand that one or the other may give better codegen in some cases
> depending on how the JIT is implemented, but having both makes it hard for
> folks to decide which one to use.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dotnet/corefx/pull/7966/files/0afdc7c9841f3c979ff9be14184c846e56aff380#r60857131>
>

jkotas replied:
Isn't there the same problem for Read? This overload is only solving half of the problem.

And it will go away with byref locals and returns in C# 7. You can just use regular assignments that are likely to produce even better codegen.

nietras replied:
Yes absolutely, `Read` has the same issue as also detailed in https://github.com/dotnet/coreclr/issues/3539 The point here is we want the **choice** between either value or ref versions since this can impact codegen and perf that is important to us. And ref locals and returns are not here yet and from @mellinoe I understand it might be months (again, even after C# 7 has been released) before `AsRef` would be added which is just too long. I'll pick that up in the converstation instead. Solving half of the problem is at least solving some of it, removing the overload would not help at all.

You could then make the same argument about `Read` versus `AsRef` why do we need `Read` if we had `AsRef` and why then even `Write`? `AsRef`would solve both presumably, but the code for this looks a lot different so in some cases we would prefer `ref` version or simple **copy** version.

jkotas replied:
What about calling this method `Copy` - so that we can make the overloads symmetric and allow solving both reading and writing of the compounded values?
```
Unsafe.Copy(ref T, void *)
Unsafe.Copy(void *, ref T)
```

jamesqo replied:
+1 to @jkotas' suggestion.

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+                                                                                                                   63 65 70 74 69 6F 6E 54 68 72 6F 77 73 01 )       // ceptionThrows.
+
+  // --- The following custom attribute is added automatically, do not uncomment -------
+  //  .custom instance void [System.Runtime]System.Diagnostics.DebuggableAttribute::.ctor(valuetype [System.Runtime]System.Diagnostics.DebuggableAttribute/DebuggingModes) = ( 01 00 02 00 00 00 00 00 ) 
```
Should this (and the comment above) be removed?

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+// Metadata version: v4.0.30319
+.assembly extern System.Runtime
+{
+  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
```
Maybe the comments in this file could be kept to 1 space after the actual code (to make it more readable)? On GitHub you have to scroll to see the end of some comments.

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+// Metadata version: v4.0.30319
```
Ditto; is this comment necessary?

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+.stackreserve 0x00100000
+.subsystem 0x0003       // WINDOWS_CUI
+.corflags 0x00000001    //  ILONLY
+// Image base: 0x02ED0000
```
Are this and the MVID comment necessary?

jkotas commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,158 @@
+        ret
+  } // end of method Unsafe::SizeOf
+
+  .method public hidebysig static void CopyBlock(void* destination, void* source, uint32 numBytesToCopy) cil managed aggressiveinlining
```
The framework convention for Copy methods is to take source first and destination second (`Array.Copy`, `Buffer.BlockCopy`, ...). Do we want to follow it here as well?

nietras commented:
> I tried to use p when the pointer wasn't really a "destination" or "source". For Cast() or InitBlock()

For `InitBlock` `p` is the destination, that is the memory that is initialized, isn't it? For the `As<T>` method using `o` instead would be fine, but `source` could be used here as well. 
`numBytesToCopy` seems a bit Win32 like so changing naming of this too based on `Buffer.MemoryCopy` which uses `sourceBytesToCopy` seems better.

> I'm happy to add extra tests to cover things that are missing. I'm not sure how we'd test this, though, any suggestions?

I am not sure either ;) I have seen tests in `coreclr` that verify the assembly generated but that might be rather tedious in this case. Perhaps some perf tests could be used to verify the calls are inlined. That is, one could compare the perf of `Unsafe` version with custom non-generic version and perhaps also a version where one makes sure the call is not inlined, the perf of these three could then be used to verify `Unsafe` calls are inlined. Perhaps there is an easier way though?

> As for adding an overload for the C#7 feature, I'm leaning towards waiting on that for now. We don't generally rely on any language features before the compiler is actually stable, and it may be a little while after that before corefx itself can actually rely on that compiler anyways. In this case, the overload wouldn't actually be usable from C# until that point.

This is my problem, "a little while after that" could again be months like adding `Unsafe` class which is too long for something as simple as this, adding it now would not take long and would make it available for experimentation too. I do understand there are many priorities here, which is why handling this now when this issue is active seems more efficient.


So to sum up I would change the API to incl. the additions I would like :):

```csharp
public static class Unsafe
{
    public static unsafe T Read<T>(void* source)
    public static unsafe void Write<T>(void* destination, T value)
    public static unsafe void Write<T>(void* destination, ref T value)
    public static unsafe int SizeOf<T>()
    
    public static T As<T>(object o)
    public static unsafe void* AsPointer<T>(ref T value)
    public static ref T AsRef<T>(void* source)
    
    public static unsafe void InitBlock(void* destination, byte initValue, uint bytesToInitialize)
    public static unsafe void CopyBlock(void* destination, void* source, uint sourceBytesToCopy)
}

public class Pinnable
{
    public byte Pin;
}
```

With regard to moving `destination` last for `InitBlock` and `CopyBlock`, would this not have a perf issue? If not, I would also move them last e.g.:

```csharp
    public static unsafe void InitBlock(byte initValue, void* destination, uint bytesToInitialize)
    public static unsafe void CopyBlock(void* source, uint sourceBytesToCopy, void* destination)
}
```

jkotas commented:
>  We don't generally rely on any language features 

This overload does not rely on C# language feature. This overload can be used from IL or other languages that support byref locals and returns (e.g. managed C++). You can even test it using reflection emit in C# if you would like.

> adding it now would not take long and would make it available for experimentation too

Agree.

mellinoe commented:
> For InitBlock p is the destination, that is the memory that is initialized, isn't it? For the As<T> method using o instead would be fine, but source could be used here as well. 

From my perspective, a `destination` parameter doesn't really make sense unless you have a `source` to contrast it with, so I wouldn't like to use it for single-address methods. Others may feel differently about this naming, though. Perhaps `startAddress` would be an okay name for `InitBlock`? I'm not sure how we feel about "address" in parameter names.

> numBytesToCopy seems a bit Win32 like so changing naming of this too based on Buffer.MemoryCopy which uses sourceBytesToCopy seems better.

I'm not sure what you mean by this (w.r.t. Win32), but I'm fine with either option. We aren't consistent in the framework with respect to names for "count" variables. `Buffer.BlockCopy` just uses `count` for the parameter, for example. Personally I'm partial to `numBytesToCopy` since it is obvious they are copied from the `source` address.

And I agree, we should put the "source" parameter before the "destination" parameter. I would prefer the "numBytes" parameter to appear last, though. It feels more natural, and `Buffer.MemoryCopy` has that parameter last, as well. I mainly have it this way because the IL instruction has the parameters in that order.

nietras commented:
> Perhaps startAddress would be an okay name for InitBlock?

I don't personally like `startAddress`, since it is also obvious, then I'd rather use `p` it also has the benefit of being short. I don't mind calling a parameter `destination` in our domain of image processing and machine learning we will often have parameters that are the destination or target of a method since memory is almost always pre-allocated and output is then a parameter to a method. In this case we will often call it `destination`.

> Buffer.BlockCopy just uses count for the parameter, for example. Personally I'm partial to numBytesToCopy since it is obvious they are copied from the source address.

I just meant that `num` is a shortening of `number` which as far as I know is not recommended in the .NET naming standard. Not sure Win32 API actually uses it but it feels like hungarian almost ;) Personally, I just don't like `number` and definitely not `num`. `count` is so much better.

Agree, `source` is obvious, perhaps instead just call it `bytesCount`. `count` in my opinion raises the question whether this is element count which might confuse some users, especially since `CopyBlock` takes a `void*` and therefore can be used with e.g. `int*` which then means users might think it uses element count. So at least `bytes` should be part of name in my opinion. `bytesCount` could then be used for `InitBlock` as well. Other options could be `bytesToCopy`, `bytesToCopyCount`

>  "source" parameter before the "destination" parameter. I would prefer the "numBytes" parameter to appear last

I am very much against changing the `CopyBlock` parameters to `void* source, void* destination, uint bytesCount` since this has the same signature as `void* destination, void* source, , uint bytesCount`. Won't most users using this be expecting `CopyBlock` to be equivalent to the `cpblk` IL instruction? Which has the latter signature. At least that would be my expectation. I'd rather keep the same signature as the IL instruction in that case, since I would assume (perhaps wrongly) that users of `Unsafe` will often know what `InitBlock` and `CopyBlock` resolve into as IL.

mellinoe commented:
@nietras , @jkotas I pushed a new version with some changes:

* I changed the names for most pointer parameters. They are now either `source` if the pointer is "read from", or `destination` if the pointer is written to. I ended up using `startAddress` for the `InitBlock` parameter because it felt most natural to me.
* I went with `byteCount` for the last parameter of `InitBlock` and `CopyBlock`.
* I've added `public static ref T AsRef<T>(void* source)`.
* Cleaned up and formatted the IL a bit better.
* Updated XML comments.

jkotas commented:
> `Write<T>(void* destination, !!T& 'value')`

This method should be called Copy, and there should be a symmetric counterpart of it with opposite types of argument.

jkotas commented:
LGTM otherwise.

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.xml:
```diff
@@ -0,0 +1,84 @@
+            </summary>
+            <param name="startAddress">The address of the start of the memory block to initialize.</param>
+            <param name="value">The value to initialize the block to.</param>
+            <param name="numBytesToCopy">The number of bytes to initialize.</param>
```
Should be `byteCount`

mellinoe replied:
Thanks, it's wrong in CopyBlock as well after my change.

jamesqo commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.xml:
```diff
@@ -0,0 +1,84 @@
+            <param name="numBytesToCopy">The number of bytes to initialize.</param>
+        </member>
+    </members>
+</doc>
```
Would it be possible to add `AsRef` to the docs, or would we have to wait until C# 7?

mellinoe replied:
It should be possible I suppose, I'm just not sure how to test it. I'll put something that I think is correct for now and we can verify it when there's some way to get Intellisense for C# 7.

mellinoe replied:
I added an entry for AsRef.

jamesqo commented:
LGTM as well, after you update the docs.

nietras commented:
> `Write<T>(void* destination, !!T& 'source')` This method should be called Copy, and there should be a symmetric counterpart of it with opposite types of argument.

As far as I recall, the original issue (how to load SIMD `Vector<>` from unsafe pointer) and following conversations discussed naming of `Read`/`Write` at length. I think most favored `Read`/`Write` I may be wrong, though. Personally, I favored `Load`/`Store` but I am content with `Read`/`Write`. At least the different names for reading and writing from/to unsafe pointer mean the readability of code using this is very clear when writing or reading from unsafe pointer, which I think is a good thing. So I would keep the `Read`/`Write` naming for the `ref` versions as well. I can live with `Copy`, though. As long as the functionality is there.

> added `public static ref T AsRef<T>(void* source)` 

Great, thanks for doing this.

LGTM, otherwise too :+1:

mellinoe commented:
> So I would keep the Read/Write naming for the ref versions as well. I can live with Copy, though. As long as the functionality is there.

I'm going to keep this as-is for now (using `Copy`). The line between reading/writing is a little blurred here since you are "reading" from one parameter and "writing" into the other, so I could see us going with either. I'm going to see if we can have a quick design meeting about some of these naming questions before we finalize this API for RTM. We may be able to rename a few things before we finally lock down.

Thanks for the discussion everyone! This should be published and usable within the next day or so (from the dev feed).

nguerrera commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,177 @@
+        .custom instance void System.Runtime.Versioning.NonVersionableAttribute::.ctor() = ( 01 00 00 00 )
+        .maxstack 1
+        ldarg.0
+        ldobj !!T
```
Looks like maxstack should be 2 here.

nguerrera commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,177 @@
+        .maxstack 2
+        ldarg.0
+        ldarg.1
+        ldobj !!T
```
Looks like maxstack should be 3 here.

nguerrera commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,177 @@
+        .maxstack 2
+        ldarg.0
+        ldarg.1
+        ldobj !!T
```
Looks like maxstack should be 3 here.

nguerrera commented:
(Disregard deleted messages about maxstack. I misread.)

MarkPflug commented:
Question about this Unsafe class: shouldn't all these methods have a valuetype (where T : struct) constraint placed on the type parameter? I believe the IL opcodes that these functions rely on only work for value types, not for reference types. Or, in the case of sizeof, will result in the size of the reference, ie either 4 or 8 depending on platform. For other cases, I think the behavior might be an exception thrown during JIT. I notice that the unit tests only cover structs.

jkotas commented:
These methods should work for object references fine. I have used in some of my code (e.g. https://github.com/dotnet/corert/blob/master/src/JitInterface/src/CorInfoImpl.cs#L46) and I have not run into any problems with it.

>  in the case of sizeof, will result in the size of the reference

Exactly like the sizeof IL instructions. These methods are meant to just expose the IL instructions - nothing more.

>  I think the behavior might be an exception thrown during JIT

If you see such exceptions, please open issues on them in CoreCLR repo - they are most likely bugs in the JIT.

MarkPflug commented:
@jkotas 
Yes, this was a misunderstanding on my part in how these opcodes work. The MSDN docs (from reflection.emit) say they operate on value types, which is wrong. They operate on values, including references.
